File: longest-increasing-subsequence/shaun_gy_bs.cpp
//grandyang dp and binary search idea, edward similar idea
//how to come up this idea and why it is correct? 
//the idea is not hard to implement though
//runtime 0ms, faster than 100%, mem less than 8%
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if (nums.empty()) return 0;
        vector<int> ends{nums[0]};
        for (auto a : nums) {
            if (a < ends[0]) ends[0] = a;
            else if (a > ends.back()) ends.push_back(a);
            else {
                //find first index in ends that >= a
                int left = 0, right = ends.size()-1;
                while (left < right) {
                    int mid = left + (right - left) / 2;
                    if (ends[mid] >= a) right = mid; //func true push to left
                    else left = mid + 1;
                    
                }
                ends[left] = a;
                //ends[right] = a; //okay  too
            }
        }
        return ends.size();
    }
};
//grandyang binary search using lower_bound
class Solution2 {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> v;
        for (auto a : nums) {
            auto it = lower_bound(v.begin(), v.end(), a);
            if (it == v.end()) v.push_back(a);
            else *it = a;
        }
        return v.size();
    }
};

//grandyang binary search using upper_bound
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
       vector<int> v;
        for (auto a : nums) {
            if (find(v.begin(), v.end(), a) != v.end()) continue;
            auto it = upper_bound(v.begin(), v.end(), a);
            if (it == v.end()) v.push_back(a);
            else *it = a;
        }
        return v.size();
    }
};

File: longest-increasing-subsequence/shaun-gy-bs.java
/*
idea forming: 
try each start, then go right and include only > numbers
time O(n^2), space O(1): only store the longest

time O(nlogn), logn may binary search

dp[i] stores the lis from 0 to i-1, sorted list
dp[i+1]= ? find nums[i] position in dp[i], then dp[i+1] = .... nums[i]
        put nums[i] at index found, O(1)
        then have to copy multiple elements, O(N)?
        s

logn is a hint to use binary search
shaun idea upon the hint and build the sorted lis with one iteration 

Runtime: 1 ms, faster than 86.31% of Java online submissions for Longest Increasing Subsequence.
Memory Usage: 37.6 MB, less than 34.00% of Java online submissions for Longest Increasing Subsequence.
*/
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> seq = new ArrayList<>();
        
        for (int x: nums) {
            if (seq.isEmpty()) seq.add(x);
            else {
                if (x < seq.get(0)) seq.set(0,x);
                else if (x > seq.get(seq.size()-1)) {
                    seq.add(x);
                }else {
                    int index = binarySearch(seq, x);
                    seq.set(index,x);
                }
            }
        }
        return seq.size();
    }
    
    private int binarySearch(List<Integer> seq, int x) {
        int left = 0, right = seq.size()-1;
        while (left < right) {
            int mid = left + (right-left)/2;
            if (seq.get(mid) >= x) {
                right = mid;
            }else { // mid < x
                left = mid+1;
            }
                
        }
        //left == right now
        return left;
    }
}
